### DISCUSSION TITLE
Maintenance Mode for Connections

### DISCUSSION THREAD

**Discussion by darwinyip** [apache/airflow#24194] [General]
**Description**

Be able to set status of connection as "being in maintenance" so that it automatically pauses all DAGs that uses it, and un-pause them once maintenance mode is over.

**Use case / motivation**

This feature will be useful when database is known to be down, so to avoid all related DAGs failing and retrying.
Also saves effort from needing someone to manually remember, pause, and un-pause DAGs when a known maintenance window is scheduled.

**Comment by coopergillan** [comment]
One way to achieve something similar with existing functionality is to use a pool. If you set up a pool for all tasks with the given connection, you can set the allotted pools to zero during maintenance. The DAGs don't even need to be paused that way, but you can still count on tasks not getting scheduled. Then when you maintenance is done you can revert the number of pools.

**Comment by eladkal** [comment]
It's a bit more complicated than what you described.
You assume here that DAG uses only one connection resource.

Consider a DAG writing to MySQL A and B.
If A is under maintance. I dont think you can assume everyone would like to pause writing to B just because A is under maintance.

**Comment by potiuk** [comment]
> It's a bit more complicated than what you described.
> You assume here that DAG uses only one connection resource.
> 
> Consider a DAG writing to MySQL A and B.
> If A is under maintance. I dont think you can assume everyone would like to pause writing to B just because A is under maintance.

I think it is an interesting idea, an it actually might be interested on Task level not DAG level. So the dags should not be paused but tasks should not be executed. 

> One way to achieve something similar with existing functionality is to use a pool

Yeah. You could. But pools are there for rather different purpose and being able to pause a connection is much more fine-grained (and still allows you to use pools for example to run tasks that require GPU).

**Comment by eladkal** [comment]
> I think it is an interesting idea, an it actually might be interested on Task level not DAG level. So the dags should not be paused but tasks should not be executed.

Maybe worth to explore the option of Active/InActive mode for Connection. Currently, when task is executed it access a Connection. The feature could also before letting task to read the connection information it will verify that the connection is Active. That way the DAG itself left untouched and if it can't access the connection it will act in a sensor like way (release worker and try reschedule later).

**Comment by mik-laj** [comment]
>  (and still allows you to use pools for example to run tasks that require GPU).

You can define multiple pools for one task.

**Comment by potiuk** [comment]
> You can define multiple pools for one task.

But is not that abuse of pools? The pool is a feature taken from celery + now it is also used for CeleryKubernetesExcecutor to distinguish "big" tasks that you can send to Kubernetes from "small" tasks that you want to send to Celery. 

Surely we can use it for a hundred other things but should we? Is it a discoverable use case? are we going to extend the documentation of pools to include that case as well? The fact that we **can** do something does not always mean that we **should** do it.

> That way the DAG itself left untouched

I think in the proposal the most important "feature" of it is (what also @eladkal wrote) that you do not have to write your DAG in a special way if you would like to drain connections. In the "pools" case you would have to remember to add specially crafted pools to every task that uses some connection (one pool per connection!) and you would have to keep somewhere a mapping -> this pool should be drained to 0 when you want to drain that connection. That's probably quite on the far side of being user-friendly.

I like the idea that when you want to drain the connection - you do something (disable? drain?) about ..... the connection and not some other abstract pool that you have to remember to add when you write the DAG.


>  if it can't access the connection it will act in a sensor like the way (release worker and try to reschedule later).

I also think the kind of "reschedule" mode is not the best idea because rather than on the DAG level, it would require to add similar logic in pretty much all tasks using the connection (check the connection, if not available - reschedule). Doing it at the scheduler time might be a far more efficient way of doing it without sending the task to the executor. And In case of Kubernetes Executor - without the overhead of creating a whole new pod, just to check the connection and shut-down with rescheduling.

I think this one has also a bit further reaching implications if we implement it. In order to do that, we must have information about the connection used by the Task. This is not very difficult, and it only requires the task to know which hooks are being used by it (assuming that each hook has a single connection). While we do not have this information now AFAIK, I think it would be worthwhile to consider adding it, because it can lead to much nicer lineage information tracking.

It can be an interesting building block and base to implement better "lineage" information in the long term. If the tasks know which connections they are using, and for each connection, they register the "resource" they are using, that's an interesting way to auto-discover the lineage information - i.e. how data passes through the whole DAG. 

For me this one looks like a rather interesting candidate for a feature for Airflow 2.1.

**Comment by mik-laj** [comment]
> But is not that abuse of pools? The pool is a feature taken from celery

Pools are used to reduce the number of tasks that access a given resource ie. limit the execution parallelism on arbitrary sets of tasks.. You can configure the value from 0 to block tasks from running, any number or -1 to disable quotas.  In my opinion, it fits perfectly here, but we do not have an automatic mechanism that will assign task with specific connections to a specific pool.

See: https://airflow.readthedocs.io/en/latest/concepts.html#pools

>  The pool is a feature taken from celery + now it is also used for CeleryKubernetesExcecutor to distinguish "big" tasks that you can send to Kubernetes from "small" tasks that you want to send to Celery.

No. CeleryExecutor and CeleryKubernetesExecutor use a queue. Queues can also be used here, but it would be an abuse and it would be very difficult to configure because the queues are configured on the target server and Airflow cannot automatically change its configuration. If a worker is assigned to queue A, then it cannot be assigned to queue B without restarting the worker. It is different with pools that are fully managed by Airflow and we have a much greater influence on their configuration and behavior.

See:  https://airflow.readthedocs.io/en/latest/executor/celery.html#queues

> The fact that we can do something does not always mean that we should do it.

I agree, but if there is one concept that has a similar application, we can think about reusing or extending it.

**Comment by mik-laj** [comment]
I checked and we don't handle multiple pools for a single task, but I think it would be worth adding. I will prepare an example in a moment, which will show how to manage connection limits using pool.

**Comment by mik-laj** [comment]
I created a PR with a cluster policy that assign pool based on connection: https://github.com/apache/airflow/pull/11033/files

**Comment by potiuk** [comment]
Nobody said it **can't** be done with pools @mik-laj. 
But I'd argue it **shouldn't** be done.

We **could** probably implement the whole scheduling logic of Airflow using cluster_policy but I bet we **shouldn't** do it.

It's a nice example to have, but I still believe it's a nice feature to have on a connection level.

**Comment by potiuk** [comment]
> I created a PR with a cluster policy that assigns pool based on connection: https://github.com/apache/airflow/pull/11033/files

I left some comments - I think the problem with this workaround is that without support from the framework, it would not work well for things like transfer operators where two hooks are used with different connection ids. So I think we should at least describe it as a 'half-baked" solution that does not work in all cases.


What I think this issue is about is to try to make sure that there is always a relation "task" <> connections no matter how it is defined and that we can "drain" the task by draining the connection. I am afraid this is a bit more than just a  simple policy to add in a more "general" case. I think currently we simply lack this relation in Airflow and I think (for the lineage usage in the future), it is a nice thing to add.

**Comment by potiuk** [comment]
> No. CeleryExecutor and CeleryKubernetesExecutor use a queue. Queues can also be used here, but it would be an abuse and it would be very difficult to configure because the queues are configured on the target server and Airflow cannot automatically change its configuration. If a worker is assigned to queue A, then it cannot be assigned to queue B without restarting the worker. It is different with pools that are fully managed by Airflow and we have a much greater influence on their configuration and behavior.

You are right, pools are indeed general feature. It is similar, but I think it's still an abuse to use one pool per connection. Pools are designed to be used by people writing DAGs  and they are far more flexible. This is like you have semaphores and mutexes in programming. Surely you can use semaphore like mutex, you even have binary semaphores, but it's still better to use mutex for locking. In our case I think this is really a binary flag on connection "active/non-active" that you can not only show in the UI but also manage from the UI next to connection. And the logic to control it is far simpler than pool logic, so why do we need to reuse it.

> I checked and we don't handle multiple pools for a single task, but I think it would be worth adding. I will prepare an example in a moment, which will show how to manage connection limits using pool.

Adding this feature would lead to quite increased complexity. What happens if one pool is exhausted and one is not. How should we deal with decreasing/increasing those? How about potential deadlocks when different tasks use different (but overlapping) sets of pools.

I think it's not worth adding multiple pools without good justification. Adding the Connection "active/inactive" flag seems like an order of magnitude simpler tasks.

**Comment by darwinyip** [comment]
I like the idea of pausing the task, but when a connection is re-enabled, it should catch up. That means that subsequent DAGs should not run until the task finishes, or at least have a mechanism for the task to catch up all the missed runs.

**Comment by potiuk** [comment]
Converted it into discussion. I do not see big need for it. But maybe some discussion can still happen there.
